exporters/inmemory/src/main/java/io/opentelemetry/exporters/inmemory/InMemoryMetricExporter.java

```diff
@@ -103,12 +105,12 @@ public void reset() {
    * <p>If this is called after {@code shutdown}, this will return {@code ResultCode.FAILURE}.
    */
   @Override
-  public ResultCode export(Collection<MetricData> metrics) {
+  public Future<ResultCode> export(Collection<MetricData> metrics) {
```

Comment:
I don't think `Future` is very useful in modern Java, since we target Java 7, we should probably have a callback parameter to the method instead.

Reply:
Java 8's `CompletableFuture` implements `Future`. `Future` describes a behaviour that is very much modern Java.

Reply:
It implement's `Future` for legacy apps :) But `Future` itself only has the possibility of calling `.get`, which is a blocking method, so it's not quite asynchronous. While we weren't really clear on when this result code would actually even be used, but if the goal is to make this fully asynchronous (no blocking for the result at all) while also returning a result, on Java 7 I think it does need to be a callback.

Reply:
`Future` is purely an interface that describes a behaviour. Its usage here is legitimate.

Reply:
I'm happy if you'd like to change it from a `Future` to a callback yourself though.

Reply:
Are there other areas of the API that prefer callbacks over what would otherwise be futures? It is important for this PR to be consistent with them of course.

Reply:
I think this is the first async API. 

If the goal is to make this a fully async API, `Future` doesn't really cut it. For example, if an event loop callback needs this result, maybe in a reactive stream, it will have to delegate to an I/O threadpool before calling `Future.get`. Perhaps this isn't really an expected use case in which case, we always expect `.get` to called in a synchronous context. If so then `Future` works fine but we have to be clear that it limits the use cases for the API.

Reply:
How would you feel about exposing Guava's `ListenableFuture` here? Back in the day, this was always my go-to before Java 8 came along. And, Guava is already on the list of dependencies.

Alternatively, OT could introduce its own `Future` subtype that could host a `ListenableFuture` and migrate to a `CompletableFuture` later on if/when the Java 7 dependency can be dropped.

WDYT?

Reply:
I like `ListenableFuture` but exposing Guava in the API might be a bit much - interested in others' thoughts.

Definitely, as opposed to a callback parameter, a custom `Future` subtype would work too. As long as it has something like `addListener`, similar to Netty's future, would work well in event loops.

Reply:
Future is no longer used. Resolving.